src/expr/src/scalar/func.rs

```diff
@@ -1713,32 +1719,43 @@ impl TimestampLike for chrono::DateTime<chrono::Utc> {
     }
 }
 
-fn date_part_interval<'a>(a: Datum<'a>, interval: Interval) -> Result<Datum<'a>, EvalError> {
+fn extract_interval(a: Datum, interval: Interval) -> Result<Datum, EvalError> {
```

Comment:
This can be nominally reduced by hoisting up the generic param's literalization up a level

```diff
diff --git a/src/expr/src/scalar/func.rs b/src/expr/src/scalar/func.rs
index 331e1acdb..3df122a53 100644
--- a/src/expr/src/scalar/func.rs
+++ b/src/expr/src/scalar/func.rs
@@ -36,7 +36,7 @@ use repr::adt::array::ArrayDimension;
 use repr::adt::datetime::{DateTimeUnits, Timezone};
 use repr::adt::interval::Interval;
 use repr::adt::jsonb::JsonbRef;
-use repr::adt::numeric::{self, DecimalLike, Numeric};
+use repr::adt::numeric::{self, Dec, DecimalLike, Numeric};
 use repr::adt::regex::Regex;
 use repr::{strconv, ColumnName, ColumnType, Datum, DatumType, Row, RowArena, ScalarType};
 
@@ -1719,18 +1719,14 @@ impl TimestampLike for chrono::DateTime<chrono::Utc> {
     }
 }
 
-fn extract_interval(a: Datum, interval: Interval) -> Result<Datum, EvalError> {
-    let units = a.unwrap_str();
-    match units.parse() {
-        Ok(units) => date_part_interval_inner::<Numeric>(units, interval),
-        Err(_) => Err(EvalError::UnknownUnits(units.to_owned())),
-    }
-}
-
-fn date_part_interval(a: Datum, interval: Interval) -> Result<Datum, EvalError> {
+fn date_part_interval<'a, T>(a: Datum<'a>, b: Datum<'a>) -> Result<Datum<'a>, EvalError>
+where
+    T: DecimalLike + Into<Datum<'static>>,
+{
     let units = a.unwrap_str();
+    let interval = b.unwrap_interval();
     match units.parse() {
-        Ok(units) => date_part_interval_inner::<f64>(units, interval),
+        Ok(units) => date_part_interval_inner::<T>(units, interval),
         Err(_) => Err(EvalError::UnknownUnits(units.to_owned())),
     }
 }
@@ -2437,7 +2433,7 @@ impl BinaryFunc {
                 ))
             }
             BinaryFunc::ExtractInterval => {
-                eager!(|a, b: Datum| extract_interval(a, b.unwrap_interval()))
+                eager!(|a, b| date_part_interval::<Numeric>(a, b))
             }
             BinaryFunc::ExtractTime => {
                 eager!(|a, b: Datum| extract_time(a, b.unwrap_time()))
@@ -2452,7 +2448,7 @@ impl BinaryFunc {
                 eager!(|a, b: Datum| extract_date(a, b.unwrap_date()))
             }
             BinaryFunc::DatePartInterval => {
-                eager!(|a, b: Datum| date_part_interval(a, b.unwrap_interval()))
+                eager!(|a, b| date_part_interval::<f64>(a, b))
             }
             BinaryFunc::DatePartTime => {
                 eager!(|a, b: Datum| date_part_time(a, b.unwrap_time()))
```

I think you can do the same thing for time, and almost the same thing for timestamp/tz; the latter just requires the existing unwrap.

Reply:
Just pushed this change.